src/v/cluster/tm_stm.cc

```diff
@@ -358,16 +363,54 @@ ss::future<> tm_stm::apply(model::record_batch b) {
       tx.id,
       tx_id);
 
+    if (tx.status == tm_transaction::tx_status::tombstone) {
+        _tx_table.erase(tx.id);
+        _tx_locks.erase(tx.id);
+        _pid_tx_id.erase(tx.pid);
+        return ss::now();
+    }
+
     auto [tx_it, inserted] = _tx_table.try_emplace(tx.id, tx);
     if (!inserted) {
         tx_it->second = tx;
     }
 
     _pid_tx_id[tx.pid] = tx.id;
 
-    expire_old_txs();
-
     return ss::now();
 }
 
+bool tm_stm::is_expired(const tm_transaction& tx) {
+    auto now_ts = model::timestamp::now();
+    return _transactional_id_expiration.count()
+           < now_ts.value() - tx.last_update_ts.value();
```

Comment:
a `timepoint - timepoint` expression returns a duration type so I don't think you need to convert everything over into scalar milliseconds via value().